src/main.cpp

```diff
@@ -6886,19 +6247,25 @@ bool SendMessages(CNode *pto)
                 // Try to find first header that our peer doesn't have, and
                 // then send all headers past that one.  If we come across any
                 // headers that aren't on chainActive, give up.
-                for (const uint256 &hash : pto->vBlockHashesToAnnounce)
+                for (const uint256 &hash : blockHashesToAnnounce)
                 {
-                    BlockMap::iterator mi = mapBlockIndex.find(hash);
-                    // BU skip blocks that we don't know about.  was: assert(mi != mapBlockIndex.end());
-                    if (mi == mapBlockIndex.end())
-                        continue;
-                    CBlockIndex *pindex = mi->second;
-                    if (chainActive[pindex->nHeight] != pindex)
+                    CBlockIndex *pindex = nullptr;
```

Comment:
Looks good, but just a thought for future optimization, we could just as well store the header rather than the hash when we have a block to announce and forgo the need for all the subsequent lookups and cs_main locks.

Reply:
good idea